File: 1679-max-number-of-k-sum-pairs/1679-max-number-of-k-sum-pairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        int res = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            if (map.getOrDefault(k - num, 0) > 0) {
                map.put(k - num, map.get(k - num) - 1);
                res++;
            } else {
                map.put(num, 1 + map.getOrDefault(num, 0));
            }
        }
        return res;
    }
}

File: 1679-max-number-of-k-sum-pairs/1679-max-number-of-k-sum-pairs.cpp
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int, int> map;
        int res = 0;
        for (int num : nums) {
            map[num]++;
        }
        
        for (auto it = map.begin(); it != map.end(); it++) {
            int num = it -> first, count = it -> second;
            if (k - num == num)
                res += count / 2;
            else if (map.count(k - num)) {
                int Min = min(count, map[k - num]);
                res += Min;
                map[num] -= Min;
                map[k - num] -= Min;
            }
        }
        return res;
    }
};